### ISSUE TITLE
Appium 1.6.4 removed “visible” attribute. xPath queries failing

### ISSUE THREAD

**Issue by freddy2** [appium/appium#8231]
Just installed appium 1.6.4 and updated all other apps with brew and npm.
Using real device with iOS 10.3.1, with latest C# appium / selenium drivers.

The product under test has pages with several buttons, all of which are enabled but only one of them is visible depending on the scenario. For Appium 1.6.3 and earlier I could do driver.FindElement(...) with an xPath "//XCUIElementTypeButton[@name='button1' and visible='true']" to make sure I could find the one that was visible. 

This no longer works. I have to remove the "and visible='true'" part go get the .FindElement to work. Without that though, the .FindElement finds all of them, not just the visible one. 

In our c# code, the IWebElement has a .Displayed property, which I can use and this does return the correct information. So it appears Appium can detect whether an element is visible or not. So how come this got removed and not backwards compatible. 

Yeah, I have a workaround. It means lots of changes and extra query time, which is already slow.

**Comment by mykola-mokhnach** [comment]
This was deliberate decision for WDA, since visibility detection for each element in the tree might produce unexpected freezes due to XCTest issues. I'd advice to try location by predicates, which supports _visible_ attribute. Also, it is times faster than xpath:

```
type == 'XCUIElementTypeButton' AND name == 'button1' AND visible == 1
```

**Comment by freddy2** [comment]
type == 'XCUIElementTypeButton' AND name == 'button1' AND visible == 1

Is there a c# example of this here driver.FindElementBy...(...)

**Comment by mykola-mokhnach** [comment]
I think you should ask this question at some other place. I personally only use java client and don't know how C# stuff is organised

**Comment by freddy2** [comment]
IWebElement node = driver.FindElement("-ios predicate string", "type=='XCUIElementTypeStaticText' AND name=='Update available' AND visible==1");

This is working for me.

**Comment by mykola-mokhnach** [comment]
You will be also able to use more advanced class chain locator with predicates, but only after WDA is merged into Appium 1.6.5-beta. Check https://github.com/facebook/WebDriverAgent/wiki/Queries for more details. "By" magic string for class chain locator is "-ios class chain"

**Comment by mykola-mokhnach** [comment]
Can the issue be closed?

**Comment by janice0402** [comment]
@freddy2 , Like this xpath : //XCUIElementTypeApplication[1]//XCUIElementTypeStaticText[contains(@name, "Item not found")], how do i write like you give the example : driver.FindElement("-ios predicate string", "type=='XCUIElementTypeStaticText' AND name=='Update available' AND visible==1");

**Comment by freddy2** [comment]
janice0402, Hi I don't know the answer to that. Yet. I want to spend time testing the changes I have made and see the test cases running again. But I do plan to investigate as there are 27 more places in our tests that use contains(@name, .....) -- I was thing try something like path=='XCUIElementTypeApplication[1]//XCUIElementTypeStaticText' for the longer more complex element part. Not sure about contains yet. Don't know of either work, have not tried it yet.
